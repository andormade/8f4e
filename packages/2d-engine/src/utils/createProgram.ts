export default function createProgram(
	gl: WebGL2RenderingContext | WebGLRenderingContext,
	shaders: WebGLShader[]
): WebGLProgram {
	const program = gl.createProgram();

	shaders.forEach(shader => {
		gl.attachShader(program, shader);
	});

	gl.linkProgram(program);

	const linked = gl.getProgramParameter(program, gl.LINK_STATUS);

	if (!linked) {
		const lastError = gl.getProgramInfoLog(program);
		console.error('Error in program linking: ' + lastError);
		gl.deleteProgram(program);
		return null;
	}

	return program;
}
