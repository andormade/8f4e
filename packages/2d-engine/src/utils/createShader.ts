export default function createShader(
	gl: WebGL2RenderingContext | WebGLRenderingContext,
	shaderSource: string,
	shaderType: number
): WebGLShader {
	const shader = gl.createShader(shaderType);
	gl.shaderSource(shader, shaderSource);
	gl.compileShader(shader);

	if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
		const info = gl.getShaderInfoLog(shader);
		console.error('Could not compile web gl shader', info);
	}

	return shader;
}
