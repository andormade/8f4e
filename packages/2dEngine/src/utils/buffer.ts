/**
 * Fills a vertex buffer with vertices of the specified rectangle.
 * @param buffer
 * @param offset
 * @param x
 * @param y
 * @param width
 * @param height
 */
export const fillBufferWithRectangleVertices = function (
	buffer: Float32Array,
	offset: number,
	x: number,
	y: number,
	width: number,
	height: number
) {
	const x1: number = x;
	const x2: number = x + width;
	const y1: number = y;
	const y2: number = y + height;

	// Triangle 1 vertex 1
	buffer[offset] = x1;
	buffer[offset + 1] = y1;
	// vertex 2
	buffer[offset + 2] = x2;
	buffer[offset + 3] = y1;
	// vertex 3
	buffer[offset + 4] = x1;
	buffer[offset + 5] = y2;

	// Triangle 2 vertex 1
	buffer[offset + 6] = x1;
	buffer[offset + 7] = y2;
	// vertex 2
	buffer[offset + 8] = x2;
	buffer[offset + 9] = y1;
	// vertex 3
	buffer[offset + 10] = x2;
	buffer[offset + 11] = y2;
};

/**
 * Fills a texture coordinate buffer with the specified sprite coordinates.
 * @param buffer
 * @param offset
 * @param spriteX
 * @param spriteY
 * @param spriteWidth
 * @param spriteHeight
 * @param spriteSheetWidth
 * @param spriteSheetHeight
 */
export const fillBufferWithSpriteCoordinates = function (
	buffer: Float32Array,
	offset: number,
	spriteX: number,
	spriteY: number,
	spriteWidth: number,
	spriteHeight: number,
	spriteSheetWidth: number,
	spriteSheetHeight: number
) {
	const u1: number = spriteX / spriteSheetWidth;
	const v2: number = (spriteY + spriteHeight) / spriteSheetHeight;
	const u2: number = (spriteX + spriteWidth) / spriteSheetWidth;
	const v1: number = spriteY / spriteSheetHeight;

	buffer[offset + 0] = u1;
	buffer[offset + 1] = v1;
	buffer[offset + 2] = u2;
	buffer[offset + 3] = v1;
	buffer[offset + 4] = u1;
	buffer[offset + 5] = v2;
	buffer[offset + 6] = u1;
	buffer[offset + 7] = v2;
	buffer[offset + 8] = u2;
	buffer[offset + 9] = v1;
	buffer[offset + 10] = u2;
	buffer[offset + 11] = v2;
};

export const fillBufferWithLineVertices = function (
	buffer: Float32Array,
	offset: number,
	x1: number,
	y1: number,
	x2: number,
	y2: number,
	thickness: number = 1
) {
	/*      o (x1, y1)
	 *      |\  Alpha
	 *      | \
	 *      |  \  <-- The hypotenuse here
	 * legA |   \     is the line we want to render.
	 *      |    \
	 *      |     \ Beta
	 *      '----- o (x2, y2)
	 *        legB
	 *
	 *  We need to calculate one of it's angles to be able
	 *  to copy and translate the original line to make it look thicker.
	 */
	const legA = y1 - y2;
	const legB = x1 - x2;
	const alpha = Math.atan(legA / legB);

	/*               |
	 *               |--- o (xA, yA)
	 *               |   /|\ Alpha
	 *               |  / | \
	 *         Beta  | /  |  \
	 *      (x1, y1) |/   |   \
	 * --------------*-)--+----\---->
	 *              /|\         \
	 *             / | \         \
	 *            /  |  \         \
	 *           /   |   \         '
	 * (xD, xD) o-)--+    \          .
	 *           \   |     \
	 *            \  |      '    <-- This is the central line
	 *             \          .      for which we have the cooridnates
	 *              '
	 *                .  <-- This is a slightly translated line
	 */

	thickness = thickness / 2;
	const translateX = Math.max(1, thickness) * Math.sin(alpha);
	const translateY = Math.max(1, thickness) * Math.cos(alpha);

	let xA = x1;
	let yA = y1;
	let xD = x1 + translateX;
	let yD = y1 - translateY;

	let xB = x2;
	let yB = y2;
	let xC = x2 + translateX;
	let yC = y2 - translateY;

	if (thickness >= 1) {
		xA = x1 - translateX;
		yA = y1 + translateY;
		xB = x2 - translateX;
		yB = y2 + translateY;
	}

	// triangle 1 vertex 1
	buffer[offset] = xA;
	buffer[offset + 1] = yA;
	// vertex 2
	buffer[offset + 2] = xB;
	buffer[offset + 3] = yB;
	// vertex 3
	buffer[offset + 4] = xC;
	buffer[offset + 5] = yC;

	// triangle 2 vertex 1
	buffer[offset + 6] = xD;
	buffer[offset + 7] = yD;
	// vertex 2
	buffer[offset + 8] = xA;
	buffer[offset + 9] = yA;
	// vertex 3
	buffer[offset + 10] = xC;
	buffer[offset + 11] = yC;
};
