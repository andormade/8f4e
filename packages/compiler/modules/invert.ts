import { i32load, i32const, i32store } from '../wasm/instructions';
import { createFunctionBody } from '../wasm/sections';
import { Instruction } from 'wasm-bytecode-utils';
import { ModuleGenerator } from '../types';

const enum Memory {
	ZERO = 0x00,
	INPUT_POINTER = 0x04,
	OUTPUT = 0x08,
}

const invert: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[],
		[
			...i32const(Memory.OUTPUT + offset),
			...i32const(Memory.INPUT_POINTER + offset),
			...i32load(),
			...i32load(),
			...i32const(-1),
			Instruction.I32_MUL, // TODO: replace MUL with bitwise operation
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		offset,
		initialMemory: [0, Memory.ZERO + offset, Memory.ZERO + offset, 0],
		memoryAddresses: [
			{ address: Memory.OUTPUT + offset, id: 'out' },
			{ address: Memory.INPUT_POINTER + offset, id: 'in', default: Memory.ZERO + offset, isInputPointer: true },
		],
	};
};

export default invert;
