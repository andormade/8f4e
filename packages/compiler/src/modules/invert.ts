import { Instruction, i32load, i32const, i32store, createFunctionBody } from '@8f4e/bytecode-utils';
import { MemoryTypes, ModuleGenerator } from '../types';

export enum Memory {
	DEFAULT_VALUE,
	INPUT_POINTER,
	OUTPUT,
}

const invert: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[],
		[
			...i32const(offset.byte(Memory.OUTPUT)),
			...i32const(offset.byte(Memory.INPUT_POINTER)),
			...i32load(),
			...i32load(),
			...i32const(-1),
			Instruction.I32_MUL, // TODO: replace MUL with bitwise operation
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		byteAddress: offset.byte(0),
		wordAddress: offset.word(0),
		memoryMap: [
			{ type: MemoryTypes.PRIVATE, address: Memory.DEFAULT_VALUE, default: 0 },

			{
				type: MemoryTypes.INPUT_POINTER,
				address: offset.byte(Memory.INPUT_POINTER),
				id: 'in',
				default: offset.byte(Memory.DEFAULT_VALUE),
			},

			{ type: MemoryTypes.OUTPUT, address: offset.byte(Memory.OUTPUT), id: 'out', default: 0 },
		],
	};
};

export default invert;
