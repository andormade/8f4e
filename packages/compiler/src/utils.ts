import { GLOBAL_ALIGNMENT_BOUNDARY } from './consts';
import { BlockStack, CompiledModule, MemoryMap, StackItem } from './types';

export function isMemoryIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return memoryMap.has(name);
}

export function isMemoryReferenceIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return (
		(name.startsWith('&') && memoryMap.has(name.substring(1))) ||
		(name.endsWith('&') && memoryMap.has(name.slice(0, -1)))
	);
}

export function isMemoryPointerIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name.startsWith('*') && memoryMap.has(name.substring(1));
}

export function isWordSpanIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name.startsWith('$') && memoryMap.has(name.substring(1));
}

export function isWordSizeIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name.startsWith('%') && memoryMap.has(name.substring(1));
}

export function getDataStructure(memoryMap: MemoryMap, id: string) {
	return memoryMap.get(id);
}

export function getDataStructureByteAddress(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.byteAddress : 0;
}

export function getMemoryStringLastAddress(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.byteAddress + (memoryItem.wordAlignedSize - 1) * GLOBAL_ALIGNMENT_BOUNDARY : 0;
}

export function getWordSpan(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.wordAlignedSize : 0;
}

export function getWordSize(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.elementWordSize : 0;
}

export function isInstructionIsInsideAModule(blockStack: BlockStack) {
	for (let i = blockStack.length - 1; i > 0; i--) {
		if (blockStack[i].isModuleBlock) {
			return true;
		}
	}
	return false;
}

export function isInstructionIsInsideAGroup(blockStack: BlockStack) {
	return blockStack[blockStack.length - 1] && blockStack[blockStack.length - 1].isGroupBlock;
}

export function calculateMemoryWordSize(memory: MemoryMap): number {
	return Array.from(memory.values()).reduce((accumulator, current) => {
		return accumulator + current.wordAlignedSize;
	}, 0);
}

export function areAllOperandsIntegers(...operands: StackItem[]): boolean {
	return !operands.some(operand => !operand.isInteger);
}

export function areAllOperandsFloats(...operands: StackItem[]): boolean {
	return !operands.some(operand => operand.isInteger);
}
