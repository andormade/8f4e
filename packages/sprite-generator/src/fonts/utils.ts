export function pad(paddingLeft: number, arr: number[], paddingRight = 0): number[] {
	return [...new Array(paddingLeft).fill(0b00000000), ...arr, ...new Array(paddingRight).fill(0b00000000)];
}

export function same(byte: number | number[], times: number): number[] {
	return new Array(times).fill(byte).flat();
}

export function mirr(toBeMirrored: number[], middle: number[] = []): number[] {
	return [...toBeMirrored, ...middle, ...toBeMirrored.slice().reverse()];
}
