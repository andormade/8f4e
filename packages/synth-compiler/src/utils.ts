import { RelativeAddressCalculator } from './types';

export function createRelativeAddressCalculator(offset: number, wordLength: number): RelativeAddressCalculator {
	return {
		byte: function (nthWord) {
			return nthWord * wordLength + offset * wordLength;
		},
		word: function (nthWord) {
			return nthWord + offset;
		},
	};
}
