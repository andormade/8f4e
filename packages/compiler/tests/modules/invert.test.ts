import { createTestModule } from '../utils';
import invert, { Memory } from '../../src/modules/invert';

let testModule;

test('if compiled module matches with snapshot', () => {
	expect(invert('id', () => 0)).toMatchSnapshot();
});

describe('functional tests', () => {
	beforeAll(async () => {
		testModule = await createTestModule(invert);
	});

	beforeEach(() => {
		testModule.reset();
	});

	test('invert module', () => {
		const { memory, test } = testModule;

		memory[Memory.DEFAULT_VALUE] = 1;
		test();
		expect(memory[Memory.OUTPUT]).toBe(-1);

		memory[Memory.DEFAULT_VALUE] = -69;
		test();
		expect(memory[Memory.OUTPUT]).toBe(69);

		memory[Memory.DEFAULT_VALUE] = 0;
		test();
		expect(memory[Memory.OUTPUT]).toBe(0);
	});
});
