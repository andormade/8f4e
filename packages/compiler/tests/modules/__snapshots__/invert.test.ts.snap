// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if compiled module matches with snapshot 1`] = `
Object {
  "functionBody": Array [
    18,
    0,
    65,
    0,
    65,
    0,
    40,
    2,
    0,
    40,
    2,
    0,
    65,
    127,
    108,
    54,
    2,
    0,
    11,
  ],
  "initialMemory": Array [
    0,
    0,
    0,
    0,
  ],
  "memoryAddresses": Array [
    Object {
      "address": 0,
      "id": "out",
    },
    Object {
      "address": 0,
      "default": 0,
      "id": "in",
    },
  ],
  "moduleId": "id",
  "offset": 0,
}
`;
